test/integration-test/src/tests/ring_buf.rs

```diff
@@ -0,0 +1,440 @@
+            .load(crate::RING_BUF)
+            .unwrap();
+        let ring_buf = bpf.take_map("RING_BUF").unwrap();
+        let ring_buf = RingBuf::try_from(ring_buf).unwrap();
+        let regs = bpf.take_map("REGISTERS").unwrap();
+        let regs = PerCpuArray::<_, Registers>::try_from(regs).unwrap();
+        let prog: &mut UProbe = bpf
+            .program_mut("ring_buf_test")
+            .unwrap()
+            .try_into()
+            .unwrap();
+        prog.load().unwrap();
+        prog.attach(
+            Some("ring_buf_trigger_ebpf_program"),
+            0,
+            "/proc/self/exe",
+            None,
+        )
+        .unwrap();
+
+        Self {
+            _bpf: bpf,
+            ring_buf,
+            regs,
+        }
+    }
+}
+
+struct WithData(RingBufTest, Vec<u64>);
+
+impl WithData {
+    fn new() -> Self {
+        Self(RingBufTest::new(), {
+            let mut rng = rand::thread_rng();
+            // Generate more entries than there is space so we can test dropping entries.
+            let n = rng.gen_range(1..=RING_BUF_MAX_ENTRIES * 2);
+            std::iter::repeat_with(|| rng.gen()).take(n).collect()
+        })
+    }
+}
+
+#[test]
+fn ring_buf() {
+    let WithData(
+        RingBufTest {
+            mut ring_buf,
+            regs,
+            _bpf,
+        },
+        data,
+    ) = WithData::new();
+
+    // Note that after expected_capacity has been submitted, reserve calls in the probe will fail
+    // and the probe will give up.
+    let expected_capacity = RING_BUF_MAX_ENTRIES - 1;
+
+    // Call the function that the uprobe is attached to with randomly generated data.
+    let mut expected = Vec::new();
+    let mut expected_rejected = 0u64;
+    let mut expected_dropped = 0u64;
+    for (i, &v) in data.iter().enumerate() {
+        ring_buf_trigger_ebpf_program(v);
+        if i >= expected_capacity {
+            expected_dropped += 1;
+        } else if v % 2 == 0 {
+            expected.push(v);
+        } else {
+            expected_rejected += 1;
+        }
+    }
+
+    let mut seen = Vec::<u64>::new();
+    while seen.len() < expected.len() {
+        if let Some(read) = ring_buf.next() {
+            let read: [u8; 8] = (*read)
+                .try_into()
+                .with_context(|| format!("data: {:?}", read.len()))
+                .unwrap();
+            let arg = u64::from_ne_bytes(read);
+            assert_eq!(arg % 2, 0, "got {arg} from probe");
+            seen.push(arg);
+        }
+    }
+
+    // Make sure that there is nothing else in the ring_buf.
+    assert_matches!(ring_buf.next(), None);
+
+    // Ensure that the data that was read matches what was passed, and the rejected count was set
+    // properly.
+    assert_eq!(seen, expected);
+    let Registers { dropped, rejected } = regs.get(&0, 0).unwrap().iter().sum();
+    assert_eq!(dropped, expected_dropped);
+    assert_eq!(rejected, expected_rejected);
```

Comment:
So this only really gets exercised some of the times. I think it should be a
separate test?

Reply:
I added two cases so that both are covered. In the "fewer than capacity" case, indeed expected_dropped is 0.